logagent.1.md
%logagent(1) user manual | version 0.0.4 85a6082
% R. S. Doiel
% 2025-03-26

# NAME

logagent

# SYNOPSIS

logagent [OPTIONS]
logagent YAML_FILE LOG_FILE [OPTIONS]

# DESCRIPTION

Log agent reads input line by line. If checks if a tag (explicit sub string)
is contained in that line. If a match is found then the agent extracts any
IP addresses identified before applying a rule associated with the tag.

The log agent requires a configuration file written in yaml. The configuration
holds an array of object where each object is an agent configuration. The
object has the following attributes.

tag
: The explicit string to search for

action
: The command to execute if tag is found

Here's an example configuration YAML file.

~~~yaml
- tag: BadBot
  action: |
    sudo iptables
    -p tcp -m multiport
    --dports http,https
    -j DROP
    -s {ipaddress}
~~~

If the text "BadBot" is found in the log line. and the IP address "156.59.198.136" was found in the log line then the following command would be executed.

~~~shell
    sudo iptables \
       -p tcp -m multiport \
       --dports http,https \
       -j DROP \
       -s 156.59.198.136
~~~

# OPTIONS

Options come as the last parameter(s) on the command line.

-h, --help
: display help

-v, --version
: display version

-l, --license
: display license

-d, --dry_run
: display the commands for matching tags in the configuration. Nice
for generating bash or Powershell scripts.

# EXAMPLES

In example we're looking for log lines that have the text "BadBot"
or "BadSpider". We'll use iptables to ban them.

Here's the YAML config called "badbots.yaml"

~~~
- tag: BadBot
  action: |
    sudo iptables
        -p tcp -m multiport
        --dports http,https
        -j DROP
        -s {ipaddress}
- tag: BadSpider
  action: |
    sudo iptables
        -I logagent_badbot
        -p tcp -m multiport
        --dports http,https
        -j DROP
        -s {ipaddress}
~~~

When you run 'logagent' with the '--dry_run' option it
will show you the commends that will be executed for log lines
with tags. Here's an example using the YAML config on "access.log"

~~~
logagent badbots.yaml access.log --dry_run
~~~

If this looks OK then you can apply the tags and actions like this.

~~~
logagent badbots.yaml access.log
~~~




logagent.ts
// logagent.ts is a TypeScript module
// By R. S. Doiel
import * as yaml from "@std/yaml";
import { parseArgs } from "@std/cli";
import { licenseText, releaseDate, releaseHash, version } from "./version.ts";
import { fmtHelp, helpTextLogAgent as helpText } from "./helptext.ts";

const reIPAddress = new RegExp('[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+');
export interface LogAgentsInterface {
  loadConfig: (arg0: string) => Promise<boolean>;
}

// LogAgent implements an object that perform action based
// on the contents of a YAML configuration file.
export class LogAgents implements LogAgentsInterface {
  agents: { tag: string; action: string }[] = [];

  // loadConfig, reads the YAML configuration and updates
  // the LogAgent's state.
  //
  // ```
  //  let la = new LogAgent();
  //  if (! await la.loadConfig("badbots.yaml")) {
  //     ... // handle failure to read and proces badbots.yaml
  //  }
  // ```
  async loadConfig(fName: string): Promise<boolean> {
    const src = await Deno.readTextFile(fName);
    if (src === undefined) {
      console.log(`failed to read ${fName}`);
      return false;
    }
    let agentList: { tag: string; action: string }[] = yaml.parse(
      src,
    ) as unknown as { tag: string; action: string }[];
    if (agentList === undefined) {
      return false;
    }
    let i = 0;
    for (let agent of agentList) {
      if (
        agent.tag === undefined || agent.action === undefined ||
        agent.tag.trim() === "" || agent.action.trim() === ""
      ) {
        console.log(
          `agent ${i} is badly formed, ${agent.toString()}, aborting load`,
        );
        return false;
      }
      agent.tag = agent.tag.trim();
      agent.action = agent.action.trim();
      this.agents.push(agent);
      i++;
    }
    return true;
  }

  async dryRun(cfgName: string, logName: string): Promise<boolean> {
    if (await this.loadConfig(cfgName)) {
        const src = await Deno.readTextFile(logName);
        let collected_actions: {[key: string]: string[]} = {};
        for (const line of src.split('\n')) {
            if (line !== '') {
                const ipaddress: string = lineToIPAddress(line);
                for (const agent of this.agents) {
                    if (line.indexOf(agent.tag) > -1) {
                      if (collected_actions[ipaddress] === undefined) {
                        collected_actions[ipaddress] = [];
                      }
                      if (collected_actions[ipaddress].indexOf(agent.action) === -1) {
                        collected_actions[ipaddress].push(agent.action);
                        console.log(`# Agent Tag: ${agent.tag}`);
                        console.log(agent.action.replaceAll('\n', ' ').replaceAll('{ipaddress}', ipaddress));
                      } 
                    }
                }    
            }
        }
        return true;
    }
    return false;
  }

  async apply(fName: string, logName: string): Promise<boolean> {
    if (await this.loadConfig(fName) === false) {
        return false;
    }
    const src = await Deno.readTextFile(logName);
    let collected_actions: {[key: string]: string[]} = {};
    
    for (const line of src.split('\n')) {
        
        const ipaddress: string = lineToIPAddress(line);
        for (const agent of this.agents) {
          let ipaddress = lineToIPAddress(line);
          if (line.indexOf(agent.tag) > -1) {
            if (collected_actions[ipaddress] === undefined) {
              collected_actions[ipaddress] = [];
            }
            if (collected_actions[ipaddress].indexOf(agent.action) === -1) {
              collected_actions[ipaddress].push(agent.action);

              let cmd = agent.action.replaceAll('{ipaddress}', ipaddress);
              console.log(cmd);
              await runCmd(cmd);
            }
          }
        }
    }
    return true;
  }
}

function lineToIPAddress(line: string): string {
    for (let part of line.split(' ')) {
        if (reIPAddress.test(part)) {
            return part;
        }
    }
    return '';
}

function parseParams(input: string): string[] {
    const params: string[] = [];
    let currentParam = '';
    let inQuotes = false;

    for (let i = 0; i < input.length; i++) {
        const char = input[i];

        if (char === '"') {
            inQuotes = !inQuotes;
        } else if (char === ' ' && !inQuotes) {
            if (currentParam.trim() !== '') {
                params.push(currentParam.trim());
                currentParam = '';
            }
        } else {
            currentParam += char;
        }
    }

    if (currentParam.trim() !== '') {
        params.push(currentParam.trim());
    }

    return params;
}
  
async function runCmd(s: string): Promise<boolean> {
    let args: string[] = parseParams(s);
    let cmd: string | undefined = args.shift();
    if (cmd === undefined) {
        return false;
    }
    const command = new Deno.Command(`${cmd}`, {
        args: args
    });
    const { code, stdout, stderr } = await command.output();
    //const out = (new TextDecoder()).decode(stdout);
    //if (out !== undefined) console.log(out);
    if (code === 0) {
        return true
    }
    const eout = (new TextDecoder()).decode(stderr);
    if (eout !== undefined) console.log(eout);
    return false;
}

async function main(): Promise<number> {
  const appName = "logagent";
  const app = parseArgs(Deno.args, {
    alias: {
      help: "h",
      license: "l",
      version: "v",
      dry_run: "d",
    },
    default: {
      help: false,
      version: false,
      license: false,
      dry_run: false,
    },
  });
  const args = app._;
  if (app.help) {
    console.log(fmtHelp(helpText, appName, version, releaseDate, releaseHash));
    Deno.exit(0);
  }
  if (app.license) {
    console.log(licenseText);
    Deno.exit(0);
  }

  if (app.version) {
    console.log(`${appName} ${version} ${releaseDate} ${releaseHash}`);
    Deno.exit(0);
  }

  const cfgName: string = (args.length > 0) ? `${args[0]}` : '';
  const logName: string = (args.length > 1) ? `${args[1]}` : '';
  if (cfgName === '' || logName === '') {
    console.log(`error: cfgName -> "${cfgName}", logName: "${logName}"`);
    Deno.exit(1);  
  }
  let agents = new LogAgents();
  if (app.dry_run) {
    if (await agents.dryRun(cfgName, logName)) {
      return 0;
    }  
  } else {
    if (await agents.apply(cfgName, logName)) {
      return 0;
    }
  }
  return 1;
}

if (import.meta.main) await main();